src/lib.rs

```diff
@@ -975,19 +1003,47 @@ impl<T> Drop for IntoIter<T> {
     }
 }
 
-/*
+impl<'a, T> Iterator for Drain<'a, T> {
+    type Item = T;
+    fn next(&mut self) -> Option<T> {
+        self.iter.next().map(|x| unsafe {
+            ptr::read(x)
+        })
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
+    }
+}
+
+impl<'a, T> DoubleEndedIterator for Drain<'a, T> {
+    fn next_back(&mut self) -> Option<T> {
+        self.iter.next_back().map(|x| unsafe {
+            ptr::read(x)
+        })
+    }
+}
+
+impl<'a, T> ExactSizeIterator for Drain<'a, T> {}
+
 impl<'a, T> Drop for Drain<'a, T> {
```

Comment:
Reading this code, it would help me if there were a comment here explaining what happens to any unconsumed items from the iterator.  (It leaks them, right?)

Reply:
Oops - it used to leak them but it shouldn't have - I've fixed that now.